homeassistant/components/withings/common.py

```diff
@@ -89,184 +479,362 @@ def __init__(
 
-        self._measures = None
-        self._sleep = None
-        self._sleep_summary = None
+    SUCCESS = 0
+    CONNECT_ERROR = 1
+    HTTP_ERROR = 2
+    NOT_WEBHOOK = 3
 
-        self.sleep_summary_last_update_parameter = None
-        self.throttle_data = {}
 
-    @property
-    def profile(self) -> str:
-        """Get the profile."""
-        return self._profile
+def async_is_valid_etag(etag: str) -> bool:
+    """Check that the etag came from the withings webhook."""
+    return etag is not None and etag.startswith(f"{const.WEBHOOK_TEST_ETAG}_")
 
-    @property
-    def slug(self) -> str:
-        """Get the slugified profile the data is for."""
-        return self._slug
 
-    @property
-    def api(self) -> ConfigEntryWithingsApi:
-        """Get the api object."""
-        return self._api
+def async_create_etag() -> str:
+    """Create a new webhook etag."""
+    return f"{const.WEBHOOK_TEST_ETAG}_{time.time()}"
 
-    @property
-    def measures(self) -> MeasureGetMeasResponse:
-        """Get the current measures data."""
-        return self._measures
 
-    @property
-    def sleep(self) -> SleepGetResponse:
-        """Get the current sleep data."""
-        return self._sleep
+async def async_test_webhook_availability(
+    hass: HomeAssistant, callback_url: str
+) -> WebhookAvailability:
+    """Test withings webhook availability."""
+    session = async_get_clientsession(hass)
+    error = None
+    error_message = None
 
-    @property
-    def sleep_summary(self) -> SleepGetSummaryResponse:
-        """Get the current sleep summary data."""
-        return self._sleep_summary
-
-    @staticmethod
-    def get_throttle_interval() -> int:
-        """Get the throttle interval."""
-        return const.THROTTLE_INTERVAL
-
-    def get_throttle_data(self, domain: str) -> ThrottleData:
-        """Get throttlel data."""
-        return self.throttle_data.get(domain)
-
-    def set_throttle_data(self, domain: str, throttle_data: ThrottleData):
-        """Set throttle data."""
-        self.throttle_data[domain] = throttle_data
-
-    @staticmethod
-    def print_service_unavailable() -> bool:
-        """Print the service is unavailable (once) to the log."""
-        if WithingsDataManager.service_available is not False:
-            _LOGGER.error("Looks like the service is not available at the moment")
-            WithingsDataManager.service_available = False
-            return True
+    try:
+        response = await session.head(
+            callback_url,
+            timeout=ClientTimeout(
+                # Withings only gives your app 2 seconds to respond.
+                total=2
+            ),
+        )
 
-        return False
+        if response.status != 200:
+            error_message = (
+                f"Received non-successful HTTP status code from {callback_url}"
+            )
+            error = WebhookAvailability.HTTP_ERROR
+
+        elif not async_is_valid_etag(response.headers.get("ETag")):
```

Comment:
Why should this webhook availability code be part of home assistant and not a 3rd party library? One of the main pillars of home assistant is to have device/service protocol specific interface code outside of home assistant.

Reply:
This webhook testing code should be part of Home Assistant because it tests Home Assistant's webhook implementation. There isn't really anything that Home Assistant is doing with the webhook that can be turned into a generic implementation.

Reply:
It's the withings api that is setting the limitations here. Home Assistant doesn't care about this.

I think this belongs in a 3rd party library.

Reply:
Withings only cares that your webhook is running on HTTPS and responds when sending data. I added this check because there are a lot of issue reports for the withings integration by people who don't understand or like the idea that they have to have home assistant available to the public during authentication.
Adding the webhook requires a URL to always be publicly accessible and I forsee numerous more tickets arriving because people disabled public access or messed up their reverse proxy config. This check helps cut down those issues by proactively informing the user about the fact their webhook URL is not accessible.

Reply:
I don't have a problem with the check as such. I just want to move the logic out of home assistant. We can call a library function from here passing the webhook url.

Reply:
I actually think that we should remove this check completely because we will run into hairpinning issues, where routers that don't support it cannot reach an external url that points at a device in their internal network.

As long as we use our new `get_url` helper and specify the right requirements (allow_interal=False), we can tell the user ahead of time if the url is going to work or not. This is a lot better experience.

Reply:
The latest commit removes this functionality.